### PR TITLE
`Ibc-rs` Documentation

### LINKED ISSUE #59: `Ibc-rs` and hyperspace documentation
Technical documentation for `ibc-rs` and `hyperspace`.

### PR DESCRIPTION
**Pull request by Wizdave97**
Partly closes #59

### CODE REVIEW

**Review by blasrodri — APPROVED** [review]
Aside from the comments on the PR on `composable` I don't have anything else to add :)

File: contracts/pallet-ibc/routing.md [hunk 1]
```diff
@@ -13,6 +13,11 @@ This means that each ibc application must statically define a unique module id a
 - Add the callback weight handler to the [`weight router`](/code/parachain/frame/ibc/src/weight.rs#L150).
 - Add the module id to the `lookup_module_by_port` implementation.
 
+**Custom Router**   
+
+Pallet ibc provides a means to use a custom router defined as a Runtime Config parameter. This sub-router implements the `ModuleRouter` trait and
```
Comment (blasrodri) [inline contracts/pallet-ibc/routing.md:18]:
```sed
s/ibc/IBC/g
```

**Review by blasrodri — DISMISSED** [review]

File: contracts/pallet-ibc/ics23.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+- `NextSequenceSend` - `"nextSequenceSend/ports/{identifier}/channels/{identifier}"`
+- `NextSequenceReceive` - `"nextSequenceRecv/ports/{identifier}/channels/{identifier}"`
+- `NextSequenceAcknowledge` - `"nextSequenceAck/ports/{identifier}/channels/{identifier}"`
+
+The approach we take to implement this is to make use of the [`child trie API`](https://github.com/paritytech/substrate/blob/master/frame/support/src/storage/child.rs), the child trie API affords us a couple benefits, listed below
+- The child trie API is a lower level storage API that allows us to insert values into our runtime storage using
+  the custom key paths provided by the IBC protocol.
+- it to allows the light client on the counterparty chain use the global state root of the host chain to verify state proofs.
+- Optimizes this pallet by removing the need to have an expensive computation that would have been incurred if we had to
+  generate a commitment root for the ibc state in this pallet's on_finalize hook
```
Comment (seunlanlege) [inline contracts/pallet-ibc/ics23.md:23]:
```suggestion

```

File: contracts/pallet-ibc/routing.md [hunk 3]
```diff
@@ -0,0 +1,288 @@
+## Routing (ICS26) and callback handlers
+
```
Comment (seunlanlege) [inline contracts/pallet-ibc/routing.md:2]: I've changed the router implementation, please look at `master` so you can update the docs

File: ibc/modules/README.md [hunk 4]
```diff
@@ -1,47 +1,212 @@
...
 
-[![Crate][crate-image]][crate-link]
-[![Docs][docs-image]][docs-link]
-[![Build Status][build-image]][build-link]
-[![End to End testing][e2e-image]][e2e-link]
 [![Apache 2.0 Licensed][license-image]][license-link]
-![Rust Stable][rustc-image]
-![Rust 1.51+][rustc-version]
 
+Implementation of the Inter-Blockchain Communication Protocol ([IBC]) in rust.
```
Comment (seunlanlege) [inline ibc/modules/README.md:5]: Definitely link the other sub docs to this one and move them all to a `docs/` subfolder

File: contracts/pallet-ibc/routing.md (hunk 3 above)
Comment (Wizdave97) [reply contracts/pallet-ibc/routing.md:2]: Looking at it, we also need a similar implementation for the benchmarks.

File: contracts/pallet-ibc/docs/routing.md [hunk 5]
```diff
@@ -175,7 +175,12 @@ impl<T: Config + Send + Sync> Module for IbcModule<T> {
         let success = "success".as_bytes().to_vec();
         let data = String::from_utf8(packet.data.clone()).ok();
         let packet = packet.clone();
```
Comment (blasrodri) [inline contracts/pallet-ibc/docs/routing.md:177]: do you actually need to clone the packet? as far as I see, it's only taking a reference in line 182

File: contracts/pallet-ibc/README.md [hunk 6]
```diff
@@ -91,7 +94,7 @@ ICS20 is an ibc application which can be implemented as a standalone pallet neve
 The core ics20 logic is already implemented in [`ibc-rs`](/ibc/modules/src/applications/transfer), all that's required to integrate this is to implement the callback handlers for ics20 
 and implement the [`Ics20Context`](/ibc/modules/src/applications/transfer/context.rs#l118) trait.
 
-This `Ics20Context` implementation is dependent on an implementation of `frame_support::traits::fungibles::{Inspect, Mutate, Transfer}` for token registration, minting, transfers and burning of tokens
+This `Ics20Context` implementation is dependent on an implementation of `frame_support::traits::fungibles::{Inspect, Mutate, Transfer}` for token registration, minting, transfers and burning.
```
Comment (blasrodri) [inline contracts/pallet-ibc/README.md:97]: This `Ics20Context` is dependent on the implementation of `frame_support::traits::fungibles::{Inspect, Mutate, Transfer}` for token registration, minting, transfers, and burning.

File: contracts/pallet-ibc/docs/benchmarks.md [hunk 7]
```diff
@@ -8,7 +8,7 @@ this is because different light clients have different header and proof verifica
 Also, all pallets plugged into ibc are required to benchmark their callbacks and
 provide a handler that implements the `CallbackWeight` trait which specifies methods that return the weight of each callback method.
 
-The benchmarking infrastructure for the [`deliver`](/contracts/pallet-ibc/src/weight.rs#L178) extrinsic defines a weight router that collects a batch of ibc messages, and calculates the total weight of processing the message batch,
+The benchmarking infrastructure for the [`deliver`](/weight.rs#L178) extrinsic defines a weight router that collects a batch of ibc messages, and calculates the total weight of processing the message batch,
```
Comment (blasrodri) [inline contracts/pallet-ibc/docs/benchmarks.md:11]: IBC

File: contracts/pallet-ibc/docs/routing.md [hunk 8]
```diff
@@ -276,13 +262,50 @@ impl<T: Config> CallbackWeight for WeightHandler<T> {
...
+// 1. Define a struct to hold the module callback handlers
+#[derive(Clone, Debug, Eq, PartialEq, Default)]
+pub struct Router {
+    // 2. Callback Handlers like this should implement `ibc::core::ics26_routing::context::Module` trait
+	pallet_example: pallet_example::IbcModule<Runtime>,
+}
+
+// 3. Implement ModuleRouter trait.
+impl ModuleRouter for Router {
+    //4. Add support for all custom routes within these function implementations
```
Comment (blasrodri) [inline contracts/pallet-ibc/docs/routing.md:286]:
```diff
-//4
+// 4
```

**Review by seunlanlege — APPROVED** [review]
Very comprehensive docs

**Pull request merged**